### ISSUE TITLE
State of `DeepModel` and `ShallowModel`.

### ISSUE THREAD

**Issue by xldenis** [xldenis/creusot#739]
I wonder if we we should re-evaluate the existence of both `DeepModel` and `ShallowModel`. Can we feasibly combine the two? Would it be excessive to get rid of `ShallowModel` entirely?

**Comment by jhjourdan** [comment]
We can discuss ShallowModel, but this is mainly a matter of notations. We don't need that, we can ask for every type which need this do declare its own model function. But this is tedious and we won't have the @ notation.

**Comment by jhjourdan** [comment]
What we can do, though, if we want to simplify the current status, which is far from ideal, is to add a trait `ShallowModelIsDeepModel` which inherits from `DeepModel`, and have a generic instance `impl<T : ShallowModelIsDeepModel> ShallowModel for T`. This would avoid the need to declare the two instances each time.

In addition, we could have derive macros that would allow creating these instances automatically for `struct`s and `enum`s.

An alternative is to have a default instance DeepModel => ShallowModel, and specialize it when we need to give a different `ShallowModel` instance. But I am a bit afraid that we might not be able to see that the dafault instance is not further specializable in practical cases, which is necessary to see its definition transparently....

**Comment by jhjourdan** [comment]
Thinking about it, I wonder about something: why do you need to define a shallow model on structs, where the model type would be a struct itself ? I can't think of a use case for this. If you need to speak about the shallow model of the fields, then you should rather do @x.field.

**Comment by jhjourdan** [comment]
Yet another option would be to define another specific notation for deep models...

**Comment by xldenis** [comment]
Il very against a second notation for deep models, I am also starting to think we should make @ postfix like verus does (avoids a bunch of parens)

**Comment by jhjourdan** [comment]
> The main reason is when you need to call a function which takes a tuple / strict which corresponds to that shallow model.

I'm trying to understand why we don't need such a thing in other tools such as Why3, or even Coq.

Why cannot you pass the concrete (non-model) value to these functions? I imagine that's what we would do in Why3...

**Comment by xldenis** [comment]
But we already have code that is generic over it: instances like ones for borrows

**Comment by jhjourdan** [comment]
> But we already have code that is generic over it: instances like ones for borrows

I had a look at all the generic code over this trait. They are of two kinds:
- Building notation instances from other notation instances. This is the case of `ShallowModel` instances for `Box`, borrows and `Ghost`, and instances of `IndexLogic` when the shallow model is `Seq`. I don't think this is problematic: this is just defining more notations.

- The definition of `SliceIndex`. This one is sealed, so not really generic...

**Comment by dewert99** [comment]
I wonder if you could use something like:
```rust
trait DeepModelForShallow {}

impl<T: DeepModelForShallow + ShallowModel> DeepModel for T
where T::ShallowModelTy: DeepModel{
    type DeepModelTy = T::ShallowModelTy::DeepModelTy;

    fn deep_model(self) -> Self::DeepModelTy {
        self.shallow_model().deep_model()
    }
}
```
I would require adding `DeepModel` implementations for the types in logic (eg. `Seq`, `Mapping` ...), but it seems like it is usually what you want, and has the nice property that `x@ == y@ ==> x.deep_model() == y.deep_model()`.

**Comment by xldenis** [comment]
There was an issue with instance overlaps when I last tried this, though I forget the specific issue I had.

**Comment by xldenis** [comment]
I was agonizing over `ShallowModel` again recently (I really hate that trait; it feels unprincipled), but I think one thing that would comfort me is if we made `x@ == y@ ==> x.deep_model() == y.deep_model()` a law of deepmodel at the very least.

**Comment by xldenis** [comment]
Part of this came up because I proved that `x@ == y@` but the code contains `assert_eq!(x, y)` !!!!! which is defined on `deep_model`..

**Comment by jhjourdan** [comment]
I don't know. It rather looks like a "good practice" than something which is required fundamentally.

**Comment by jhjourdan** [comment]
Ok, right, I see the argument.

But then, that would mean that we require a ShallowModel for any type implementing DeepModel. Do we really want this?

**Comment by xldenis** [comment]
> But then, that would mean that we require a ShallowModel for any type implementing DeepModel. Do we really want this?

Not necessarily: we can word the law as such

```
#[requires(x@ == y@)] #[ensures(x.deep_model() == y.deep_model())]
fn shallow_implies_deep(x : Self, y : Self) where Self : ShallowModel;

**Comment by jhjourdan** [comment]
Oh wow! Is that sound?

We probably already discussed that, but that's a bit frightening because we rely on the orphan rules to make sure that no instance is going to appear in a different crate....

**Comment by jhjourdan** [comment]
If a type has `DeepModel` but not `ShallowModel`, then the proof obligation will not be required. But if in a later crate we implement `ShallowModel`, the proof obligation will not show up either.

**Comment by xldenis** [comment]
But a later crate *can't* implement it.

**Comment by jhjourdan** [comment]
That's my point: we are using orphan rules to guarantee soundness. That's a bit frightening to me, because orphan rules are designed for coherence, which is different to guarantee that such and such trait impl does not exist.

**Comment by xldenis** [comment]
Coherence is important for the soundness of certain Rust apis as far as I know as well, I don't think this meaningfully changes things.

**Comment by xldenis** [comment]
A thought which occured to me recently was that if we introduce `#[pure]` it would become possible to remove `DeepModel` entirely. We could then specify `ParitalEq` directly in terms of `eq` being an equivalence relation.

However, @jhjourdan pointed out that this could introduce issues for the solvers as they could find themselves required to prove that `eq` is a morphism for different operations (ie `PartialOrd`). I still think that the benefit is high enough that we should attempt the experiment when possible.

**Comment by dewert99** [comment]
I seem like `DeepModel` is necessary to match how Creusot currently specifies `T: PartialEq<U>` since it forces `T` `U` to have the same `DeepModelTy` and guarantees that equality is transitive across types

**Comment by xldenis** [comment]
You're right that we would only be able to specify homogenous equality :/ I think in practice this covers 99.9% of uses of `PartialEq` however.

**Comment by dewert99** [comment]
I wonder if it would still be worth having the `DeepModel` for specifying the `Borrow` trait.
The current documentation for `Borrow` states:
> Further, when providing implementations for additional traits, it needs to be considered whether they should behave identically to those of the underlying type as a consequence of acting as a representation of that underlying type. Generic code typically uses Borrow<T> when it relies on the identical behavior of these additional trait implementations. These traits will likely appear as additional trait bounds.
> 
> In particular Eq, Ord and Hash must be equivalent for borrowed and owned values: x.borrow() == y.borrow() should give the same result as x == y.

Which could currently be handled using the specification
```rust
pub trait Borrow<Borrowed>
where
    Borrowed: ?Sized,
{
    #[ensures(result.deep_model() == self.deep_model())]
    fn borrow(&self) -> &Borrowed
    where
        Self: DeepModel,
        Borrowed: DeepModel<DeepModelTy = Self::DeepModelTy>;;
}

**Comment by xldenis** [comment]
Perhaps, my big issue with `DeepModel` is the burden it places on users, its primarily bureaucratic overhead for your Rust types. The only 'interesting' instances are for collection types like `HashMap`.

**Comment by jhjourdan** [comment]
Another issue we should be cautious about this approach is that it may require that pure functions are compatible with program equality, which is probably not what we want. For example, a function returning the capacity of a vector is a fine logic and program funstion, and thus a fine pure function, but it is not compatible with program equality, which ignores vector capacity.

**Comment by jhjourdan** [comment]
That being said, your idea gives me another: we could keep `DeepModel` (and call it differently, say `EqModel`), but provide another way to implement it (via another trait) with an equivalence relation, which is more practical to implement. An axiomatized generic instance would provide an abstract `DeepModelTy` and a projection to this type.

**Comment by jhjourdan** [comment]
This is, anyway, I think, the most efficient way to tell an SMT solver that something is an equivalence reltion and that some functions are morphisms for it.
